source/common/protobuf/utility.cc

```diff
@@ -83,43 +88,63 @@ ProtoValidationException::ProtoValidationException(const std::string& validation
   ENVOY_LOG_MISC(debug, "Proto validation error; throwing {}", what());
 }
 
-ProtoUnknownFieldsMode MessageUtil::proto_unknown_fields = ProtoUnknownFieldsMode::Strict;
-
-void MessageUtil::loadFromJson(const std::string& json, Protobuf::Message& message) {
-  MessageUtil::loadFromJsonEx(json, message, ProtoUnknownFieldsMode::Strict);
-}
-
-void MessageUtil::loadFromJsonEx(const std::string& json, Protobuf::Message& message,
-                                 ProtoUnknownFieldsMode proto_unknown_fields) {
+void MessageUtil::loadFromJson(const std::string& json, Protobuf::Message& message,
+                               ProtobufMessage::ValidationVisitor& validation_visitor) {
   Protobuf::util::JsonParseOptions options;
-  if (proto_unknown_fields == ProtoUnknownFieldsMode::Allow) {
-    options.ignore_unknown_fields = true;
-  }
   options.case_insensitive_enum_parsing = true;
-  const auto status = Protobuf::util::JsonStringToMessage(json, &message, options);
-  if (!status.ok()) {
-    throw EnvoyException("Unable to parse JSON as proto (" + status.ToString() + "): " + json);
+  // Let's first try and get a clean parse when checking for unknown fields;
+  // this should be the common case.
+  options.ignore_unknown_fields = false;
+  const auto strict_status = Protobuf::util::JsonStringToMessage(json, &message, options);
+  if (strict_status.ok()) {
+    // Success, no need to do any extra work.
+    return;
   }
+  // If we fail, we see if we get a clean parse when allowing unknown fields.
+  // This is essentially a workaround
+  // for://github.com/protocolbuffers/protobuf/issues/5967.
+  // TODO(htuch): clean this up when protobuf supports JSON/YAML unknown field
+  // detection directly.
+  options.ignore_unknown_fields = true;
+  const auto relaxed_status = Protobuf::util::JsonStringToMessage(json, &message, options);
```

Comment:
do you think there are any perf concerns for doing this parse twice? I think not, but wanted to double check. If we care about perf you could probably short circuit by asking the visitor whether it wants to accept unknown fields or not which I think would be a very simple interface change. Not sure if that is worth it or not.

Reply:
I think this won't have any performance implications on the happy path, since we only do a single JSON -> protobuf conversion there. It's only when we fail that we double work, in which case we are rejecting config. Once we have a clean solution to https://github.com/protocolbuffers/protobuf/issues/5967, even the unhappy path can be fast.

Reply:
Sounds good.